### PR TITLE
JSON::Pure fix strict mode

### PR DESCRIPTION
**Pull request by casperisfine**
Followup: https://github.com/flori/json/pull/519
Fix: https://github.com/flori/json/issues/584

### CODE REVIEW

File: lib/json/pure/generator.rb [hunk 1]
```diff
@@ -354,7 +354,7 @@ def json_transform(state)
               result << delim unless first
               result << state.indent * depth if indent
               result = "#{result}#{key.to_s.to_json(state)}#{state.space_before}:#{state.space}"
-              if state.strict?
+              if state.strict? && !(!value || value == true || value.is_a?(Array) || value.is_a?(Integer) || value.is_a?(Float) || value.is_a?(String))
```
Comment (eregon) [inline lib/json/pure/generator.rb:357]: This looks expensive, can't we just rely on `.to_json(state)` calling e.g. Object#to_json and that will raise?

**Comment by eregon** [comment]
I also noticed the docs in https://github.com/flori/json/blob/4f876a8bf7c45972800bae5bdbaf892aa2f714a5/lib/json/pure/generator.rb#L222 are wrong, could you fix that too?

And the alignment in https://github.com/flori/json/blob/4f876a8bf7c45972800bae5bdbaf892aa2f714a5/lib/json/pure/generator.rb#L143 is inconsistent (I dislike this kind of alignment, too high maintenance and diff noise, but might as well be consistent in one way, i.e. all or none aligned)

File: lib/json/pure/generator.rb (hunk 1 above)
Comment (casperisfine) [reply lib/json/pure/generator.rb:357]: I don't think we can as it wouldn't catch objects with a custom `to_json`.

**Comment by casperisfine** [comment]
> I also noticed

I'd rather not bundle multiple unrelated fixes in the same PR. Makes merging more complicated.

**Comment by eregon** [comment]
Ah sorry, it's `@script_safe` misaligned, and that's unrelated indeed.

OTOH I think fixing the description of `def strict?` is very related.

File: lib/json/pure/generator.rb (hunk 1 above)
Comment (eregon) [reply lib/json/pure/generator.rb:357]: Re `!value` that could be an issue if someone redefines `!` on their type to return true in some cases.
Not sure whether it's worth caring about, redefining `!` seems a pretty poor idea (maybe it should be a Ruby warning even).

**Review by eregon — APPROVED** [review]
Thanks for the fix!

**Pull request merged**